README.rst
BitShares proxy
---------------

This is a simple script that acts as an RPC proxy to the BitShares client
and provides simple access control to it, allowing to restrict access to
certain methods of the client only in a flexible way.

If you like this tool, please vote for `my
delegate <http://digitalgaia.io/btstools.html>`__ to support further
development, and feel free to visit my page for other delegate proposals
at `digitalgaia.io <http://digitalgaia.io>`__. Thanks!


Install
=======

To install, run::

    $ pip install bts_proxy

preferably (but not required) in a virtualenv.


Running the proxy
=================

Just run the ``bts-proxy`` script on the command line.

Configuration file with listening port and users/passwords alongside with
allowed methods can be found in the same data dir as the one for the BitShares
client. Config file is named ``proxy.json``, and will be created automatically
the first time you run ``bts-proxy`` if it doesn't exist yet.

You can specify a different data dir than the default one for the BitShares client
as the first argument to the ``bts-proxy`` script, e.g.::

    $ bts-proxy ~/.BitShares


Configuration file format
=========================

The configuration file is a simple JSON file, looking like this::

    {
        "port": 5681,

        "users": [
            {
                "name": "username",
                "password": "secret-password",
                "methods_allowed": ["*"]
            }
        ]
    }

``port`` is the port number on which the proxy will be listening. Make sure
to use a different port than the one on which the BitShares client is listening!

``users`` is a list of objects containing the following fields:

* ``name``: the name of the user
* ``password``: the password for that user
* ``methods_allowed``: the list of methods allowed. You can used shell-like
  pattern matching here (eg: ``"wallet_*"`` will allow all methods starting
  with ``"wallet_"``)
* ``methods_forbidden`` *(optional)*: the list of forbidden methods. You can also use
  shell-like pattern matching here.

Note that by default, methods are forbidden, so the proxy will allow you to
call a method **if and only if** it appears in the ``methods_allowed`` field
and **not** in the ``methods_forbidden`` field.

Security best practices
=======================

It is highly recommended to run both the BitShares client and the RPC proxy as
their own separate user, rather than your common one, and to restrict read access to
the BitShares client data dir to only this user.

This should come by default with BitShares >= 0.5.0, otherwise you can do the
following::

    $ chmod 700 ~/.BitShares


HISTORY.rst
.. This is your project NEWS file which will contain the release notes.
.. Example: http://www.python.org/download/releases/2.6/NEWS.txt
.. The content of this file, along with README.rst, will appear in your
.. project's PyPI page.

History
=======

0.1 (2015-01-10)
----------------

* first public release


bts_proxy/proxy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# bts_proxy - Proxy providing RPC access control to the BitShares client
# Copyright (c) 2015 Nicolas Wack <wackou@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

from http.server import BaseHTTPRequestHandler, HTTPServer
from os.path import expanduser, join, exists
import argparse
import base64
import requests
import fnmatch
import json
import sys
import logging

log = logging.getLogger(__name__)

PORT = None
USER = None
PASS = None

CONFIG = None


class UnauthorizedError(Exception):
    pass


def default_data_dir():
    if sys.platform.startswith('linux'):
        return '~/.BitShares'
    elif sys.platform == 'darwin':
        return '~/Library/Application Support/BitShares'
    else:
        return None


DEFAULT_CONFIG = """
{
    "port": 5681,

    "users": [
        {
            "name": "username",
            "password": "secret-password",
            "methods_allowed": ["*"]
        }
    ]
}
"""


def load_config(data_dir=None):
    global CONFIG, PORT, USER, PASS

    data_dir = data_dir or default_data_dir()
    if not data_dir:
        print('You need to specify a data dir for the proxy to fetch RPC information!')
        print('Exiting...')
        sys.exit(1)

    data_dir = expanduser(data_dir)
    config_file = join(data_dir, 'proxy.json')

    if not exists(config_file):
        print('Could not find config file for proxy...')
        print('Creating default one in: %s' % config_file)
        print('Please edit and configure with your own values')
        with open(config_file, 'w') as f:
            f.write(DEFAULT_CONFIG)
        sys.exit(1)

    with open(config_file) as f:
        CONFIG = json.load(f)

    # TODO: validate config schema/values

    # get RPC connection info to client
    try:
        with open(join(data_dir, 'config.json')) as f:
            cfg = json.load(f)['rpc']
            PORT = int(cfg['httpd_endpoint'].split(':')[1])
            USER = cfg['rpc_user']
            PASS = cfg['rpc_password']
    except Exception:
        print('Failed to read bts client RPC config:')
        raise


def rpc_call(data):
    host = 'localhost'
    url = "http://%s:%d/rpc" % (host, PORT)
    headers = {'Content-Type': 'application/json'}

    response = requests.post(url,
                             auth=(USER, PASS),
                             data=data,
                             headers=headers,
                             stream=True)

    if response.status_code == 401:
        raise UnauthorizedError

    return response.raw.read()


class Handler(BaseHTTPRequestHandler):
    def do_HEAD(self):
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()

    def do_AUTHHEAD(self):
        #print('send header AUTHHEAD')
        self.send_response(401)
        self.send_header('WWW-Authenticate', 'Basic realm="bts wallet"')
        self.send_header('Content-Type', 'text/html')
        self.end_headers()

    def do_POST(self):
        try:
            return self._do_POST()
        except Exception as e:
            log.exception(e)
            self.send_response(500)
            self.send_header('Content-Type', 'text/html')
            self.end_headers()
            self.wfile.write(b'Internal server error')
            return


    def _do_POST(self):
        ''' Present frontpage with user authentication. '''
        header = self.headers.get('Authorization')

        if header is None:
            self.do_AUTHHEAD()
            self.wfile.write(b'no auth header received')
            return

        header = header.encode('utf-8')

        for user in CONFIG['users']:
            auth_string = '%s:%s' % (user['name'], user['password'])
            auth_string = base64.b64encode(auth_string.encode('utf-8'))
            valid_auth_header = b' '.join([b'Basic', auth_string])

            if header == valid_auth_header:
                # properly authenticated
                length = int(self.headers['Content-Length'])
                request = self.rfile.read(length).decode('utf-8')

                # filter allowed methods
                payload = json.loads(request)
                method = payload['method']

                allowed = False
                for pattern in user.get('methods_allowed', []):
                    if fnmatch.fnmatch(method, pattern):
                        allowed = True
                        break
                for pattern in user.get('methods_forbidden', []):
                    if fnmatch.fnmatch(method, pattern):
                        allowed = False
                        break

                if not allowed:
                    self.send_response(403)
                    self.send_header('Content-Type', 'text/html')
                    self.end_headers()
                    msg = 'Forbidden: user "%s" is not authorized to call method "%s"' % (user['name'], method)
                    self.wfile.write(msg.encode('utf-8'))
                    return

                # forward rpc call to client
                try:
                    result = rpc_call(request)
                    self.do_HEAD()
                    self.wfile.write(result)
                    return

                except UnauthorizedError:
                    self.do_AUTHHEAD()
                    self.wfile.write(b'Unauthorized connection from proxy to BitShares client')
                    return

                except requests.exceptions.ConnectionError:
                    self.send_response(504)
                    self.send_header('Content-Type', 'text/html')
                    self.end_headers()
                    self.wfile.write(b'Could not connect to BitShares client')
                    return

        # wrong authentication (no corresponding user/password)
        self.do_AUTHHEAD()
        self.wfile.write(b'Unauthorized')



def main():
    parser = argparse.ArgumentParser(description='Run a filtering proxy for the BitShares client.')
    parser.add_argument('data_dir', metavar='data_dir', type=str, nargs='?',
                       help='the data dir of the BitShares client')

    args = parser.parse_args()

    load_config(data_dir=args.data_dir)

    print('Starting proxy server on port: %d' % CONFIG['port'])
    httpd = HTTPServer(('', CONFIG['port']), Handler)

    httpd.serve_forever()


bts_proxy/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# bts_proxy - Proxy providing RPC access control to the BitShares client
# Copyright (c) 2015 Nicolas Wack <wackou@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#


bts_proxy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# bts_proxy - Proxy providing RPC access control to the BitShares client
# Copyright (c) 2014 Nicolas Wack <wackou@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

from http.server import BaseHTTPRequestHandler, HTTPServer
from os.path import expanduser, join, exists
import argparse
import base64
import requests
import fnmatch
import json
import sys
import logging

log = logging.getLogger(__name__)

PORT = None
USER = None
PASS = None

CONFIG = None


class UnauthorizedError(Exception):
    pass


def default_data_dir():
    if sys.platform.startswith('linux'):
        return '~/.BitShares'
    elif sys.platform == 'darwin':
        return '~/Library/Application Support/BitShares'
    else:
        return None


DEFAULT_CONFIG = """
{
    "port": 5681,

    "users": [
        {
            "name": "username",
            "password": "secret-password",
            "methods_allowed": ["*"]
        }
    ]
}
"""


def load_config(data_dir=None):
    global CONFIG, PORT, USER, PASS

    data_dir = data_dir or default_data_dir()
    if not data_dir:
        print('You need to specify a data dir for the proxy to fetch RPC information!')
        print('Exiting...')
        sys.exit(1)

    data_dir = expanduser(data_dir)
    config_file = join(data_dir, 'proxy.json')

    if not exists(config_file):
        print('Could not find config file for proxy...')
        print('Creating default one in: %s' % config_file)
        print('Please edit and configure with your own values')
        with open(config_file, 'w') as f:
            f.write(DEFAULT_CONFIG)
        sys.exit(1)

    with open(config_file) as f:
        CONFIG = json.load(f)

    # TODO: validate config schema/values

    # get RPC connection info to client
    try:
        with open(join(data_dir, 'config.json')) as f:
            cfg = json.load(f)['rpc']
            PORT = int(cfg['httpd_endpoint'].split(':')[1])
            USER = cfg['rpc_user']
            PASS = cfg['rpc_password']
    except Exception:
        print('Failed to read bts client RPC config:')
        raise


def rpc_call(data):
    host = 'localhost'
    url = "http://%s:%d/rpc" % (host, PORT)
    headers = {'Content-Type': 'application/json'}

    response = requests.post(url,
                             auth=(USER, PASS),
                             data=data,
                             headers=headers,
                             stream=True)

    if response.status_code == 401:
        raise UnauthorizedError

    return response.raw.read()


class Handler(BaseHTTPRequestHandler):
    def do_HEAD(self):
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()

    def do_AUTHHEAD(self):
        #print('send header AUTHHEAD')
        self.send_response(401)
        self.send_header('WWW-Authenticate', 'Basic realm="bts wallet"')
        self.send_header('Content-Type', 'text/html')
        self.end_headers()

    def do_POST(self):
        try:
            return self._do_POST()
        except Exception as e:
            log.exception(e)
            self.send_response(500)
            self.send_header('Content-Type', 'text/html')
            self.end_headers()
            self.wfile.write(b'Internal server error')
            return


    def _do_POST(self):
        ''' Present frontpage with user authentication. '''
        header = self.headers.get('Authorization')

        if header is None:
            self.do_AUTHHEAD()
            self.wfile.write(b'no auth header received')
            return

        header = header.encode('utf-8')

        for user in CONFIG['users']:
            auth_string = '%s:%s' % (user['name'], user['password'])
            auth_string = base64.b64encode(auth_string.encode('utf-8'))
            valid_auth_header = b' '.join([b'Basic', auth_string])

            if header == valid_auth_header:
                # properly authenticated
                length = int(self.headers['Content-Length'])
                request = self.rfile.read(length).decode('utf-8')

                # filter allowed methods
                payload = json.loads(request)
                method = payload['method']

                allowed = False
                for pattern in user.get('methods_allowed', []):
                    if fnmatch.fnmatch(method, pattern):
                        allowed = True
                        break
                for pattern in user.get('methods_forbidden', []):
                    if fnmatch.fnmatch(method, pattern):
                        allowed = False
                        break

                if not allowed:
                    self.send_response(403)
                    self.send_header('Content-Type', 'text/html')
                    self.end_headers()
                    msg = 'Forbidden: user "%s" is not authorized to call method "%s"' % (user['name'], method)
                    self.wfile.write(msg.encode('utf-8'))
                    return

                # forward rpc call to client
                try:
                    result = rpc_call(request)
                    self.do_HEAD()
                    self.wfile.write(result)
                    return

                except UnauthorizedError:
                    self.do_AUTHHEAD()
                    self.wfile.write(b'Unauthorized connection from proxy to BitShares client')
                    return

                except requests.exceptions.ConnectionError:
                    self.send_response(504)
                    self.send_header('Content-Type', 'text/html')
                    self.end_headers()
                    self.wfile.write(b'Could not connect to BitShares client')
                    return

        # wrong authentication (no corresponding user/password)
        self.do_AUTHHEAD()
        self.wfile.write(b'Unauthorized')



def main(args):
    load_config(data_dir=args.data_dir)

    print('Starting proxy server on port: %d' % CONFIG['port'])
    httpd = HTTPServer(('', CONFIG['port']), Handler)

    httpd.serve_forever()

if __name__ == '__main__':

    parser = argparse.ArgumentParser(description='Run a filtering proxy for the BitShares client.')
    parser.add_argument('data_dir', metavar='data_dir', type=str, nargs='?',
                       help='the data dir of the BitShares client')

    args = parser.parse_args()
    main(args)
